src/lib/Preview.js

```diff
@@ -479,13 +485,29 @@ class Preview extends EventEmitter {
     download() {
         const { apiHost, queryParams } = this.options;
 
-        if (checkPermission(this.file, PERMISSION_DOWNLOAD)) {
-            // Append optional query params
-            const downloadUrl = appendQueryParams(getDownloadURL(this.file.id, apiHost), queryParams);
-            get(downloadUrl, this.getRequestHeaders()).then((data) => {
-                openUrlInsideIframe(data.download_url);
-            });
+        if (!checkPermission(this.file, PERMISSION_DOWNLOAD)) {
+            return;
         }
+
+        // Append optional query params
+        const downloadUrl = appendQueryParams(getDownloadURL(this.file.id, apiHost), queryParams);
+        get(downloadUrl, this.getRequestHeaders()).then((data) => {
+            const defaultDownloadUrl = replaceDownloadHostWithDefault(data.download_url);
+            if (isDownloadHostBlocked() || !isCustomDownloadHost(data.download_url)) {
+                // If we know the host is blocked, or we are already using the default,
+                // use the default.
+                openUrlInsideIframe(defaultDownloadUrl);
+            } else {
+                // Try the custom host, then check reachability
+                openUrlInsideIframe(data.download_url);
+                setDownloadReachability(data.download_url).then(() => {
+                    if (isDownloadHostBlocked()) {
+                        // If download is unreachable, try again with default
+                        openUrlInsideIframe(defaultDownloadUrl);
```

Comment:
Is there any opportunity for a double download here? e.g. can the first download succeed but then this reachability check fail?

Reply:
I don't believe so. The only way the reachability check fails is if the network request itself fails. This is different from any bad status `500`, `404` etc. which will be treated as "successes" because we get a valid HTTP status.